libfwupdplugin/fu-udev-device.c

```diff
@@ -518,6 +521,124 @@ fu_udev_device_get_slot_depth (FuUdevDevice *self, const gchar *subsystem)
 	return 0;
 }
 
+/**
+ * fu_udev_device_unbind_module:
+ * @self: A #FuUdevDevice
+ * @error: A #GError, or %NULL
+ *
+ * Unbinds the module from the device, which normally means the kernel driver
+ * releases the hardware and usuallly means it stops working.
+ *
+ * If there is no driver bound then this function will return with success
+ * without actually doing anything.
+ *
+ * Returns: %TRUE if driver was unbound.
+ *
+ * Since: 1.5.0
+ **/
+gboolean
+fu_udev_device_unbind_module (FuUdevDevice *self, GError **error)
+{
+#ifndef _WIN32
+	FuUdevDevicePrivate *priv = GET_PRIVATE (self);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GOutputStream) stream = NULL;
+
+	/* is already unbound */
+	fn = g_build_filename (g_udev_device_get_sysfs_path (priv->udev_device),
+			       "driver", NULL);
+	if (!g_file_test (fn, G_FILE_TEST_EXISTS))
+		return TRUE;
+
+	/* write bus ID to file */
+	if (priv->bus_id == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "bus-id not set for subsystem %s",
+			     priv->subsystem);
+		return FALSE;
+	}
+	file = g_file_new_for_path (fn);
+	stream = G_OUTPUT_STREAM (g_file_replace (file, NULL, FALSE,
+				  G_FILE_CREATE_NONE, NULL, error));
+	if (stream == NULL)
+		return FALSE;
+	return g_output_stream_write_all (stream, priv->bus_id, strlen (priv->bus_id),
+					  NULL, NULL, error);
+#else
+	g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "driver unbinding not supported on Windows");
+	return FALSE;
+#endif
+}
+
+/**
+ * fu_udev_device_bind_module:
+ * @self: A #FuUdevDevice
+ * @subsystem: A subsystem string, e.g. `pci`
+ * @driver: A module name, e.g. `tg3`
+ * @error: A #GError, or %NULL
+ *
+ * Binds the module to the device, which normally means the kernel driver takes
+ * over the hardware and means it will continue to work as before.
+ *
+ * Returns: %TRUE if driver was bound.
+ *
+ * Since: 1.5.0
+ **/
+gboolean
+fu_udev_device_bind_module (FuUdevDevice *self,
+			    const gchar *subsystem,
+			    const gchar *driver,
+			    GError **error)
+{
+#ifndef _WIN32
+	FuUdevDevicePrivate *priv = GET_PRIVATE (self);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GOutputStream) stream = NULL;
+
+	/* driver exists */
+	fn = g_strdup_printf ("/sys/module/%s/drivers/%s:%s/bind",
```

Comment:
Is this path correct when complied into the kernel?

Reply:
Yes, that works too. Is that a hint that you think it should be `s/fu_udev_device_bind_module/fu_udev_device_bind_driver`?

Reply:
Yes, you're binding a device to a driver not to a module.

Reply:
You're completely right. I've fixed up the naming and the comments.